janusgraph-core/src/main/java/org/janusgraph/graphdb/query/ResultSetIterator.java

```diff
@@ -36,7 +37,7 @@
 
 
     public ResultSetIterator(Iterator<R> inner, int limit) {
```

Comment:
I don't have a strong opinion on this but if the `ResultSetIterator` itself is a `CloseableIterator`, would it make sense to enforce it's inner iterator to be closeable, too?

Reply:
I actually indeed tried that at first, but I ended up changing a lot more classes, and then I gave up and kept it like this. Not because that is undoable, but because I want to keep initial changes simple. If people think that is the better way to go, I'll do that and amend this commit.

Reply:
Understood. As I said, I'm also happy with your solution but if someone has a different oppinion, please tell :smile:

Reply:
I made them auto-closeable as well. To correctly delegate Traversal.close(), I ended up refactoring/modifying many iterator classes.

Reply:
@rngcntr Could be great if you could describe your thoughts a little bit, why should we force to accept `CloseableIterator` and not just `Iterator`. I'm still trying to review this PR, I guess I may miss some ideas / code style / patterns / etc. begind this logic. Also, I posted my thoughts here: https://github.com/JanusGraph/janusgraph/pull/2218#discussion_r537111253 . I may totally be wrong because I didn't research it and didn't spend much time reviewing this PR but for some reason, I feel that we can achieve better results if we don't enforce `CloseableIterator` to be an input and just force `CloseableIterator` to be be automatically closed when the last element is read.

Going in Javadoc to `java.io.Closeable`, it says:  
> Closes this stream and releases any system resources associated with it. If the stream is already closed then invoking this method has no effect.

We do implement safe `close` method in `SubqueryIterator` and this method is automatically called when we read last element from `elementIterator`, so, I'm not sure why we are forcing `CloseableIterator` to be used everywhere even so it should be used just in a couple of places. 
Could be great to hear some thoughts about it.